internal/controller/clusterinstance/rendered_object.go

```diff
@@ -96,19 +108,55 @@ func (r *RenderedObject) GetResourceId() string {
 	return GetResourceId(r.GetName(), r.GetNamespace(), r.GetKind())
 }
 
-func (s *syncWaveMap) getAscendingSyncWaves() []int {
+func (r *RenderedObject) ManifestReference() *v1alpha1.ManifestReference {
+	syncWave, err := r.GetSyncWave()
+	if err != nil {
+		return nil
+	}
+	return &v1alpha1.ManifestReference{
+		APIGroup:        ptr.To(r.GetAPIVersion()),
+		Kind:            r.GetKind(),
+		Name:            r.GetName(),
+		Namespace:       r.GetNamespace(),
+		SyncWave:        syncWave,
+		LastAppliedTime: metav1.Now(),
+	}
+}
+
+// MatchesIdentity checks if two RenderedObjects are equal based on identifying fields.
+// These fields are APIGroup, Kind, Name, and Namespace.
+func (r *RenderedObject) MatchesIdentity(other *RenderedObject) bool {
+	return r.GetAPIVersion() == other.GetAPIVersion() &&
+		r.GetKind() == other.GetKind() &&
+		r.GetName() == other.GetName() &&
+		r.GetNamespace() == other.GetNamespace()
+}
+
+// IndexOfObjectByIdentity searches for a RenderedObject in the given RenderedObject slice based on
+// identity fields and returns its index.
+// It returns -1 if the target is not found.
+func IndexOfObjectByIdentity(target *RenderedObject, objects []RenderedObject) int {
+	for i, obj := range objects {
+		if obj.MatchesIdentity(target) {
+			return i
+		}
+	}
+	return -1
+}
+
+func (s *SyncWaveMap) GetAscendingSyncWaves() []int {
 	syncWaves := maps.Keys(s.data)
 	sort.Ints(syncWaves)
 	return syncWaves
 }
 
-func (s *syncWaveMap) getDescendingSyncWaves() []int {
+func (s *SyncWaveMap) GetDescendingSyncWaves() []int {
 	syncWaves := maps.Keys(s.data)
 	sort.Sort(sort.Reverse(sort.IntSlice(syncWaves)))
 	return syncWaves
 }
 
-func (s *syncWaveMap) getSlice(syncWave int) []RenderedObject {
+func (s *SyncWaveMap) GetSlice(syncWave int) []RenderedObject {
```

Comment:
Now that this is a published API it might need a better name. Why not `GetObjectsForSyncWave`?